tolstoy/src/tx_client.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::BTreeMap;
+use std::collections::btree_map::Entry;
+
+use rusqlite;
+
+use errors::{
+    Result,
+    ErrorKind
+};
+
+use mentat_db::types::{
+    Entid
+};
+
+use mentat_db::{
+    entids,
+    TypedSQLValue
+};
+
+use mentat_core::{
+    TypedValue
+};
+
+#[derive(Debug)]
+pub struct TxPart {
+    pub e: Entid,
+    pub a: i64,
+    pub v: TypedValue,
+    pub added: i32
+}
+
+#[derive(Debug)]
+pub struct Tx {
+    pub tx: Entid,
+    pub tx_instant: TypedValue,
+    pub parts: Vec<TxPart>
+}
+
+struct RawTx {
+    e: Entid,
+    a: i64,
+    v: TypedValue,
+    tx: Entid,
+    added: i32
+}
+
+pub trait TxReader {
+    fn all(sqlite: &rusqlite::Connection) -> Result<Vec<Tx>>;
+}
+
+pub struct TxClient {}
+
+impl TxReader for TxClient {
+    fn all(sqlite: &rusqlite::Connection) -> Result<Vec<Tx>> {
+        // Make sure a=txInstant rows are first, so that we process
+        // all transactions before we process any transaction parts.
```

Comment:
SQLite will be much happier with you if you don't do this. You're using a computed value for sorting, so it'll read the entire table into memory in order to sort it before it gives you a cursor.

The only index on `transactions` is `(tx, added)`, so try to make all of your queries on it start with `WHERE tx …` — e.g., "give me txes after tx1234" — or have no `WHERE` or `ORDER BY` at all.